### DISCUSSION TITLE
Option for generating complete bundle without parsing through source code

### DISCUSSION THREAD

**Discussion by VaZark** [tailwindlabs/tailwindcss#11857] [Help] [answered]
Based on the docs, it seems like we HAVE to parse through the source code to generate an production build. The playCDN is recommended only for dev.

Is there a an option to generate the full bundle (without scanning our source code) with config only for theme and plugins? I’m willing to tradeoff not being able to use dynamic values in my classes.

I plan on using the tailwind cli to dynamically generate the css based on just some config during runtime for providing runtime client-level theming for a SaaS product. 

So not having to use a docker instance with (/ pull)  the entire source code would be a decent optimisation considering how minimal tailwind is.

**Comment by wongjn** [comment]
If you really wanted every single class possible, you would need to generate every possible combination of the variants, (like `dark:first:hover:focus:md:visited:last:only:…:m-1`) so there would be all the "base" classes plus all these permutations of variants and thus the total number of possible classes would be something like (70![^1] + 1) × number of "base" classes. This assumes variant order does not matter but actually it does for some variants so we'd need to do something like (70! ÷ (70 - 1)!) + (70! ÷ (70 - 2)!) + (70! ÷ (70 - 3)!) + … +  (70! ÷ (70 - 69)!) + 70!. The "real" number would be slightly less since some variants don't make sense together, i.e. `sm:md:lg:xl:m-1` but that would be up to you to work out which combinations should be discarded.

That being said to do this, you'd want to use [a super-broad regular expression `safelist` with the full list of modifiers and combinations](https://tailwindcss.com/docs/content-configuration#using-regular-expressions):

```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  // …
  safelist: [
    'text-2xl',
    'text-3xl',
    {
      pattern: /^/,
      variants: [
        'sm',
        'md',
        'sm:md',
        'lg',
        'sm:md:lg',
        // Etc, probably have some algorithm to generate this list.
      ],
    },
  ],
  // …
}
```

[^1]: 70 factorial, equal to 11978571669969891796072783721689098736458938142546425857555362864628009582789845319680000000000000000

**Reply by VaZark** [reply]
Hahahaha. I love this take. “You’d be shooting yourselves in the foot, but here’s your bullet”

fair enough. Are css variables fair game in the theme config? I could have a base css generated by tailwind during build and have a a second css file that overrides just the css variables for colours

**Reply by wongjn** [reply, marked as answer]
Yes, you can use CSS variables in the Tailwind configuration. Consider reviewing [recommendations for CSS variable usage with colors in the documentation](https://tailwindcss.com/docs/customizing-colors#using-css-variables).

**Reply by VaZark** [reply]
That checked out perfectly. Thanks a bunch!
